src/runtime_src/hip/api/hip_module.cpp

```diff
@@ -1,64 +1,81 @@
 // SPDX-License-Identifier: Apache-2.0
 // Copyright (C) 2023-2024 Advanced Micro Device, Inc. All rights reserved.
 
-#include "core/common/error.h"
-
-#include "hip/config.h"
-#include "hip/hip_runtime_api.h"
-
+#include "hip/core/common.h"
 #include "hip/core/module.h"
 
 namespace xrt::core::hip {
 static void
 hip_module_launch_kernel(hipFunction_t f, uint32_t gridDimX, uint32_t gridDimY,
-                      uint32_t gridDimZ, uint32_t blockDimX, uint32_t blockDimY,
-                      uint32_t blockDimZ, uint32_t sharedMemBytes, hipStream_t hStream,
-                      void** kernelParams, void** extra)
+                         uint32_t gridDimZ, uint32_t blockDimX, uint32_t blockDimY,
+                         uint32_t blockDimZ, uint32_t sharedMemBytes, hipStream_t hStream,
+                         void** kernelParams, void** extra)
 {
-  if (!f)
-    throw xrt_core::system_error(hipErrorInvalidResourceHandle, "function is nullptr");
+  throw_invalid_resource_if(!f, "function is nullptr");
+
+  auto func_hdl = reinterpret_cast<function_handle>(f);
+  auto hip_mod = module_cache.get(static_cast<function*>(func_hdl)->get_module());
+  throw_invalid_resource_if(!hip_mod, "module associated with function is unloaded");
+
+  auto hip_func = hip_mod->get_function(func_hdl);
+  throw_invalid_resource_if(!hip_func, "invalid function passed");
 
   throw std::runtime_error("Not implemented");
 }
 
-static hipFunction_t
+static function_handle
 hip_module_get_function(hipModule_t hmod, const char* name)
 {
-  if (!name || strlen(name) == 0)
-    throw xrt_core::system_error(hipErrorInvalidValue, "name is invalid");
+  throw_invalid_value_if((!name || strlen(name) == 0), "name is invalid");
 
-  if (!hmod)
-    throw xrt_core::system_error(hipErrorInvalidResourceHandle, "module is nullptr");
+  throw_invalid_resource_if(!hmod, "module is nullptr");
 
-  throw std::runtime_error("Not implemented");
+  auto mod_hdl = reinterpret_cast<module_handle>(hmod);
+  auto hip_mod = module_cache.get(mod_hdl);
+  throw_invalid_resource_if(!hip_mod, "module not available");
+
+  auto hip_func = std::make_shared<function>(mod_hdl, std::string(name));
+  auto func_hdl = hip_func.get();
+  hip_mod->add_function(func_hdl, std::move(hip_func));
+  return func_hdl;
```

Comment:
Perhaps a terser API could be
```suggestion
  return  hip_mod->add_function(std::make_shared<function>(mod_hdl, std::string(name));
```

Reply:
@keryell thanks for the suggestion, I have made changes. Please review